foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/internal/helper/ConcurrencyUtil.java

```diff
@@ -0,0 +1,1490 @@
+                    stackTraceStringToStackTraceExampleNumber.put(stackTraceInformation, stackTraceId);
+                    writer.write(TraceLocalization.buildMessage("concurrency_util_summary_read_locks_on_thread_step002_5", new Object[] {readLockNr, stackTraceId, stackTraceInformation}));
+                }
+                writer.write("\n\n");
+            }
+        }
+
+        // (d) We have some more information to pump out namely errors we have traced each time the number of readers was decremented
+        writer.write("\n\n");
+        writer.write(TraceLocalization.buildMessage("concurrency_util_summary_read_locks_on_thread_step002_6", new Object[] {threadName, readLockManager.getRemoveReadLockProblemsDetected().size()}));
+        long releaseReadLockProblemNr = 0;
+        for (String currentReleaseReadLockProblem : readLockManager.getRemoveReadLockProblemsDetected()) {
+            releaseReadLockProblemNr++;
+            writer.write(TraceLocalization.buildMessage("concurrency_util_summary_read_locks_on_thread_step002_7", new Object[] {releaseReadLockProblemNr, currentReleaseReadLockProblem}));
+        }
+        writer.write("\n\n");
+        return writer.toString();
+    }
+
+    private static String getPropertyConcurrencyManagerAcquireWaitTime() {
+        return (PrivilegedAccessHelper.shouldUsePrivilegedAccess()) ?
+                AccessController.doPrivileged(new PrivilegedGetSystemProperty(SystemProperties.CONCURRENCY_MANAGER_ACQUIRE_WAIT_TIME))
+                : System.getProperty(SystemProperties.CONCURRENCY_MANAGER_ACQUIRE_WAIT_TIME, "0");
+    }
+
+    /**
+     * This helper API is created due to the problem of the corruption of the eclipselink cache. The idea is to have a
+     * tool that allows us to know specifically where the current thread was located when it acquired a READ LOCK.
+     *
+     * <P>
+     * Cache corruption problem: <br>
+     * namely the fact that when dead locks are seen to be taking place some of the threads that seem to be primary
+     * culprits of the dead lock are actually idle doing nothing but they have have left the number of readers of the
+     * cache corrupted (e.g. typically forever incremnted).
+     *
+     * @return get the stack trace of the current thread.
+     */
+    private String enrichGenerateThreadDumpForCurrentThread() {
+        final Thread currentThread = Thread.currentThread();
+        final long currentThreadId = currentThread.getId();
+
+        try {
+            // (a) search for the stack trace of the current
+            final ThreadMXBean threadMXBean = ManagementFactory.getThreadMXBean();
+            final ThreadInfo[] threadInfos = threadMXBean.getThreadInfo(new long[] { currentThreadId }, 700);
+            for (ThreadInfo threadInfo : threadInfos) {
+                final StringWriter writer = new StringWriter();
+                enrichGenerateThreadDumpForThreadInfo(writer, threadInfo);
+                return writer.toString();
```

Comment:
This looks strange - asking for an array of threads  while passing in just one id and looping through the result expecting to handle first item only. Isn't different [`ThreadMXBean.getThreadInfo()`](https://docs.oracle.com/javase/8/docs/api/java/lang/management/ThreadMXBean.html#getThreadInfo-long-int-) what you want?